aptos-move/dynamic-transaction-composer/src/builder.rs

```diff
@@ -0,0 +1,593 @@
+            },
+        };
+
+        if &ty != expected_ty {
+            bail!("Type mismatch when passing arugments around");
+        }
+        Ok(())
+    }
+
+    pub fn add_batched_call(
+        &mut self,
+        module: String,
+        function: String,
+        ty_args: Vec<String>,
+        args: Vec<Argument>,
+    ) -> anyhow::Result<Vec<Argument>> {
+        let ty_args = ty_args
+            .iter()
+            .map(|s| TypeTag::from_str(s))
+            .collect::<anyhow::Result<Vec<_>>>()?;
+        let module = ModuleId::from_str(&module)?;
+        let function = Identifier::new(function)?;
+
+        let target_module = match self.modules.get(&module) {
+            Some(module) => module,
+            None => {
+                bail!("Module {} is not yet loaded", module);
+            },
+        };
+
+        let call_idx = self
+            .builder
+            .import_call_by_name(function.as_ident_str(), target_module)?;
+        let call_type_params = ty_args
+            .iter()
+            .map(|ty| import_type_tag(&mut self.builder, ty, &self.modules))
+            .collect::<PartialVMResult<Vec<_>>>()?;
+
+        let subst_mapping = call_type_params
+            .iter()
+            .enumerate()
+            .map(|(idx, ty)| (idx as u16, ty.clone()))
+            .collect::<BTreeMap<_, _>>();
+
+        let mut arguments = vec![];
+        let expected_args_ty = {
+            let script = self.builder.as_script();
+            let func = script.function_handle_at(call_idx);
+            if script.signature_at(func.parameters).0.len() != args.len() {
+                bail!(
+                    "Function {}::{} argument call size mismatch",
+                    module,
+                    function
+                );
+            }
+            script
+                .signature_at(func.parameters)
+                .0
+                .iter()
+                .map(|ty| ty.instantiate(&subst_mapping))
+                .collect::<Vec<_>>()
+        };
+
+        for (arg, ty) in args.into_iter().zip(expected_args_ty) {
+            match arg {
+                Argument::Local(i) => {
+                    self.check_argument_compatibility(&i, &ty)?;
+                    arguments.push(i)
+                },
+                Argument::Signer(i) => arguments.push(AllocatedLocal {
+                    op_type: ArgumentOperation::Copy,
+                    is_parameter: true,
+                    local_idx: i,
+                }),
+                Argument::Raw(bytes) => {
+                    let new_local_idx = self.parameters_ty.len() as u16;
+                    self.parameters_ty.push(ty);
+                    self.parameters.push(bytes);
+                    arguments.push(AllocatedLocal {
+                        op_type: ArgumentOperation::Move,
+                        is_parameter: true,
+                        local_idx: new_local_idx,
+                    })
+                },
```

Comment:
Is it necessary to allocate a local for this argument?

Reply:
Another related question: do we worry about running out of locals? There are only 256 of them IIRC

Reply:
IMO there are two ways to do it:
1. Import the argument into constant pool
2. Assign a parameter location and pass it as argument to the script.

I would personally prefer 2 because it makes script caching a bit easier, as the script body part would remain the same if there's a list of commonly composed script.

On running out of locals, this is indeed a concern. There are only 256 of them and we need it for # of signers + # of parameters + # of returns. I don't want to over optimize for it rightaway but I don't see a clear path getting rid of the constraint rightway either.

Reply:
> I would personally prefer 2 because it makes script caching a bit easier, as the script body part would remain the same if there's a list of commonly composed script.

So all `Argument::Raw` are exposed as script parameters? If so, I find the idea both intriguing and worrying. It's good that we may be able to cache more stuff, but it also feels like a very leaky abstraction -- users may be surprised by the extra parameters the builder inserted.

> There are only 256 of them and we need it for # of signers + # of parameters + # of returns.

I was more thinking about returns. Wondering if we could make more use of the stack. Agreed it's not necessarily something we need to optimize at this moment though.

> I don't want to over optimize for it rightaway but I don't see a clear path getting rid of the constraint rightway either.

Relaxing it at the file format/vm level is probably not that hard (but still a nasty chore). @wrwg just curious, would the bytecode verifier have trouble with, let's say, 1024 locals?